### PR TITLE
(mostly) Server side  protocol checks improved

### PR DESCRIPTION
**Pull request by akallabeth**

### CODE REVIEW

File: libfreerdp/core/gcc.c [hunk 1]
```diff
@@ -681,6 +778,90 @@ BOOL gcc_write_user_data_header(wStream* s, UINT16 type, UINT16 length)
...
+}
+
+static BOOL updateEarlyServerCaps(rdpSettings* settings, UINT32 earlyCapabilityFlags,
+                                  UINT32 connectionType)
+{
+	WINPR_ASSERT(settings);
+
+	settings->SupportDynamicTimeZone =
+	    settings->SupportDynamicTimeZone && (earlyCapabilityFlags & RNS_UD_SC_DYNAMIC_DST_SUPPORTED)
+	        ? TRUE
```
Comment (hardening) [inline libfreerdp/core/gcc.c:859]: The tri-operator is not needed, as the result is already a bool. Or `settings->SupportDynamicTimeZone && !!(earlyCapabilityFlags & RNS_UD_SC_DYNAMIC_DST_SUPPORTED)` does the job ?

File: libfreerdp/core/gcc.c [hunk 2]
```diff
@@ -681,6 +778,90 @@ BOOL gcc_write_user_data_header(wStream* s, UINT16 type, UINT16 length)
...
+{
+	WINPR_ASSERT(settings);
+
+	if (settings->SupportMonitorLayoutPdu)
+		settings->SupportMonitorLayoutPdu =
+		    earlyCapabilityFlags & RNS_UD_CS_SUPPORT_MONITOR_LAYOUT_PDU;
+
+	if (settings->SupportHeartbeatPdu)
+		settings->SupportHeartbeatPdu =
+		    (earlyCapabilityFlags & RNS_UD_CS_SUPPORT_HEARTBEAT_PDU) ? TRUE : FALSE;
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply libfreerdp/core/gcc.c:827]: @hardening why `!!`? the `TRUE : FALSE` is there for readability, the datatype is already `BOOL` a `typedef` to `stdbool` where possible) and `!!` is quite cryptic (double invert to get from `int` to `bool` values)

[the remark this answers is not in the document]
Comment (hardening) [reply libfreerdp/core/gcc.c:827]: Well quite a lot of projects use `!!`

File: libfreerdp/core/server.h [hunk 3]
```diff
@@ -66,6 +66,8 @@ struct rdp_peer_channel
 	BYTE dvc_open_state;
 	UINT32 dvc_total_length;
 	rdpMcsChannel* mcsChannel;
+
+	char name[128];
```
Comment (hardening) [inline libfreerdp/core/server.h:70]: Perhaps it should be `channelName` to match other fields and we can probably move it with other `channelX` fields as I'm not sure we have much to care about ABI here.

Comment (akallabeth) [reply libfreerdp/core/server.h:70]: good idea

File: libfreerdp/core/server.h [hunk 4]
```diff
@@ -66,6 +66,8 @@ struct rdp_peer_channel
 	BYTE dvc_open_state;
 	UINT32 dvc_total_length;
 	rdpMcsChannel* mcsChannel;
+
+	char channelName[128];
```
Comment (hardening) [inline libfreerdp/core/server.h:70]: Just wondering, if I understand correctly this is a static channel, perhaps we can reduce the array size as the name is limited to 8 or 16 (can't remember) ?

File: winpr/libwinpr/utils/collections/PubSub.c [hunk 5]
```diff
@@ -185,7 +185,9 @@ int PubSub_OnEvent(wPubSub* pubSub, const char* EventName, void* context, const
 	size_t index;
 	wEventType* event;
 	int status = -1;
-	WINPR_ASSERT(pubSub);
+
+	if (!pubSub)
```
Comment (hardening) [inline winpr/libwinpr/utils/collections/PubSub.c:189]: Do you have a case in mind where the assert was not appropriate (it's public API so yeah we shall not assert, but just wondering) ?

**Review by hardening — APPROVED** [review]

File: libfreerdp/core/server.h (hunk 4 above)
Comment (akallabeth) [reply libfreerdp/core/server.h:70]: no, this is also the dynamic channel struct. did not want to add fancy `malloc` / `free` handling just to have a debug name

**Pull request merged**